AWS: Fix security group races

justinsb opened the pull request:
Fix two unlikely but hit-in-the-real-world (during conformance tests) race conditions in AWS, around eventual consistency in the AWS EC2 API for security groups.  Essentially retry a sensible number of times in the hope we eventually get a consistent view.

justinsb committed:
AWS: Fix race-condition in tagging of security group

When we create a security-group in the AWS API, there is sometimes
a delay before we can tag it (the AWS API is eventually consistent).

So we wrap CreateTags in a simple retry loop.

justinsb committed:
AWS: Fix race in security-group read/create

We need to find the ID for a named security group, or create a new one.

We do this by listing the security groups, and then doing a create if we
cannot find one.  This is a race though; against another thread if the
AWS API were consistent, but generally because the AWS API is actually
eventually consistent.

We wrap it in a retry loop.

justinsb commented:
Not sure if anyone else has ever seen this.  I might have just got really unlucky.  We should merge in >=v1, but if I'm the only person ever to have hit this I don't think I can justify merging in ==v1

justinsb commented:
So this happened for someone else: #11324.

Maybe we should merge, though I don't think I can justify including it in 1.0.  But TBH I'm a little confused ... merging to master targets e.g. 1.0.1, but only things we cherry-pick onto release-1.0 will be in 1.0?  Is that correct?

thockin commented:
We are intentionally keeping the gate shut to master until we're done
stabilizing 1.0.0, probably resulting in 1.0.1.  This sounds like it is not
urgent for a cherry pick THIS WEEK (where stability is paramount).  But you
can make your case if you want.

On Wed, Jul 15, 2015 at 1:27 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> So this happened for someone else: #11324
> <https://github.com/GoogleCloudPlatform/kubernetes/issues/11324>.
>
> Maybe we should merge, though I don't think I can justify including it in
> 1.0. But TBH I'm a little confused ... merging to master targets e.g.
> 1.0.1, but only things we cherry-pick onto release-1.0 will be in 1.0? Is
> that correct?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/11029#issuecomment-121736900>
> .
>

justinsb commented:
I agree that it should wait at this point.  The problem happens, but rarely.  The consequences are potentially pretty bad right now (#11324), but I think the real fix for that is to add a delay before retrying on failure, as this is just one way the API call could fail.

iterion commented:
@justinsb I believe I ran into this issue as well. I resolved it by manually dropping a security group in AWS, from that point the ELB was created correctly. I've only seen it happen once so far, but does seem to be a recurring issue.

bhouse commented:
I'm not sure if this error relates to this PR, but I'm having problems with ELB creation tagging the security group:

```
Events:
  FirstSeen				LastSeen			Count	From			SubobjectPath	Reason				Message
  Wed, 22 Jul 2015 17:54:38 +0000	Wed, 22 Jul 2015 17:54:38 +0000	1	{service-controller }			creating loadbalancer failed	failed to create external load balancer for service default/truth-frontend: error tagging security group: InvalidParameterValue: Value ( '' ) for parameter Tag key is invalid. Tag key cannot be null or empty.
					status code: 400, request id: []
```

The security group gets created, but there are no tags, and it doesn't proceed to creating the ELB

brendandburns commented on pkg/cloudprovider/aws/aws.go:
```diff
@@ -1656,37 +1662,54 @@ func (s *AWSCloud) removeSecurityGroupIngress(securityGroupId string, removePerm
+		createResponse, err := s.ec2.CreateSecurityGroup(createRequest)
+		if err != nil {
+			ignore := false
+			if awsError, ok := err.(awserr.Error); ok {
```
I think this is cleaner as:

```go
switch err.(type) {
case awserr.Error:
   ...
default:
  return "", err
}

justinsb replied:
Well, I still need the cast, but I'll try it the more go-way.

brendandburns replied:
Actually, you don't if you do:
```go
switch err := err.(type) {
...
}
```

Golang will do the right type case for you into `err`

brendandburns commented on pkg/cloudprovider/aws/aws.go:
```diff
@@ -1702,7 +1725,7 @@ func (s *AWSCloud) ensureSecurityGroup(name string, description string, vpcID st
 	tagRequest.Resources = []*string{&groupID}
 	tagRequest.Tags = tags
-	_, err = s.ec2.CreateTags(tagRequest)
+	_, err := s.createTags(tagRequest)
```
```go
if _, err := s.createTags(tagRequest); err != nil {
...
```

(compact this line and the next)

brendandburns commented on pkg/cloudprovider/aws/aws.go:
```diff
@@ -1710,6 +1733,31 @@ func (s *AWSCloud) ensureSecurityGroup(name string, description string, vpcID st
+// We retry mainly because if we create an object, we cannot tag it until it is "fully created" (eventual consistency)
+// The error code varies though (depending on what we are tagging), so we simply retry on all errors
+func (s *AWSCloud) createTags(request *ec2.CreateTagsInput) (*ec2.CreateTagsOutput, error) {
+	attempt := 0
```
Optional (potentially) cleaner approach to this style of thing:

```go
tick := time.Tick(1 * time.Second)
for {
   select {
   case <- tick:
     ...
   case <- time.After(60 * time.Second):
   ...
}
```

You can do this if you want, definitely not required.

justinsb replied:
I really should be doing exponential backoff here I think.  AWS had a pretty bad outage in the early days because everyone was just doing straight loops, and so a degraded API grew into a full on outage, as I recall.  I hope we can implement that using a reusable Go function but I'm not entirely sure how best to express that.  Maybe something like time.Tick but with a timeout and backoff?   I'll add a TODO to add exponential backoff in the meantime.

brendandburns commented:
Minor comments.

justinsb committed:
Fixes per review

Primarily go style issues; also a TODO that really exponential backoff
is the correct policy for API call retries.

brendandburns commented on pkg/cloudprovider/aws/aws.go:
```diff
@@ -1656,37 +1662,56 @@ func (s *AWSCloud) removeSecurityGroupIngress(securityGroupId string, removePerm
+		createResponse, err := s.ec2.CreateSecurityGroup(createRequest)
+		if err != nil {
+			ignore := false
+			switch err.(type) {
```
Sorry, I wasn't fully clear in my earlier example:

```
switch err := err.(type) {
  case awserr.Error:
     if err.Code() == "InvalidGroup.Duplicate" ...
...
```

should work.

brendandburns commented:
Shippable looks like a flake.  hitting it again.

brendandburns commented:
Marking LGTM to make progress.  @justinsb you can do the golang clean-up as a follow on.

evanphx commented:
Is there a reason this wasn't in 1.0.3?

mikedanese commented:
Likely no one ever submitted a cherry pick request. I don't think it was intentionally left out.

https://github.com/kubernetes/kubernetes/blob/master/docs/devel/cherry-picks.md

evanphx commented:
@justinsb I'm not a contributor otherwise I'd submit it. Can you?

roberthbailey commented:
@evanphx For future reference, anyone can send a cherry pick request -- it's a PR generated by `./hack/cherry_pick_pull.sh` against the release-1.0 branch.
